app/gui2/src/util/data/vec2.ts

```diff
@@ -11,12 +11,26 @@ export class Vec2 {
 
   static Zero: Vec2
 
-  static FromArr(arr: [number, number]): Vec2 {
-    return new Vec2(arr[0], arr[1])
+  static FromXY(point: Readonly<{ x: number; y: number }>): Vec2 {
+    return new Vec2(point.x, point.y)
   }
 
-  static FromDomPoint(point: DOMPoint): Vec2 {
-    return new Vec2(point.x, point.y)
+  static FromSize(point: Readonly<{ width: number; height: number }>): Vec2 {
+    return new Vec2(point.width, point.height)
+  }
+
+  static FromClientSize(point: Readonly<{ clientWidth: number; clientHeight: number }>): Vec2 {
+    return new Vec2(point.clientWidth, point.clientHeight)
+  }
+
+  static DotProduct(...values: Vec2[]): Vec2 {
+    let x = 1
+    let y = 1
+    for (const value of values) {
+      x *= value.x
+      y *= value.y
+    }
+    return new Vec2(x, y)
```

Comment:
This is not a dot product, since it produces a vector. This appears to be a Hadamard (component-wise) product instead.